src/Core/PolynomialFunction.java

```diff
@@ -129,31 +132,98 @@ public void derive() {
+        } catch (InterruptedException | ExecutionException e) {
+            e.printStackTrace();
+        }
+    }
+
+    private static final double EPSILON = 1E-15;
+
+    private Set<Double> calcRoots(double start, double end, double step) {
+        Set<Double> roots = new LinkedHashSet<>();
+        double prevY = evaluate(start).y;
         for (double x = start + step; x <= end; x += step) {
-            Vector2D point = evaluate(x);
-            if (prevPoint.y * point.y <= 0) {
-                roots.add(x - step / 2);
+            double y = evaluate(x).y;
+            // If the value is incredibly close to zero, it is a root => Add
+            if (Math.abs(y) < EPSILON) {
+                double root = formatDouble(x);
+                if (shouldAddValue(root, roots)) {
+                    roots.add(root);
+                }
+            // First use Incremental Search to check for a sign change (if-statement)
+            // then calculate an approximately really close position using linear interpolation
+            } else if (prevY * y < 0) {
+                double root = (x - step) + (x - (x - step)) * Math.abs(prevY) / (Math.abs(prevY) + Math.abs(y));
+                root = formatDouble(root);
+                if (shouldAddValue(root, roots)) {
+                    roots.add(root);
+                }
             }
-            prevPoint = point;
+            prevY = y;
         }
+        return roots;
     }
 
-    public void calcExtremePoints(double start, double end, double step) {
-        extremePoints = new ArrayList<>();
-        Vector2D startPoint = evaluate(start);
-        Vector2D nextPoint = evaluate(start + step);
-        double prevSlope = (nextPoint.y - startPoint.y) / step;
-        for (double x = start + 2 * step; x <= end; x += step) {
-            Vector2D point = evaluate(x);
-            double slope = (point.y - evaluate(x - step).y) / step;
-            if (prevSlope * slope <= 0) {
-                extremePoints.add(x - step);
+    private Set<Double> calcExtremes(double start, double end, double step) {
+        Set<Double> extremes = new LinkedHashSet<>();
+        double prevSlope = evaluateDerivative(start);
+        for (double x = start + step; x <= end; x += step) {
+            double slope = evaluateDerivative(x);
+            // If the value of the derivative is incredibly close to zero, it is an extreme point => Add
+            if (Math.abs(slope) < EPSILON) {
+                double extreme = formatDouble(x);
+                if (shouldAddValue(extreme, extremes)) {
+                    extremes.add(extreme);
+                }
+            // First use Incremental Search to check for a sign change (if-statement)
+            // then calculate an approximately really close position using linear interpolation
+            } else if (prevSlope * slope < 0) {
+                double extreme = (x - step) + (x - (x - step)) * Math.abs(prevSlope) / (Math.abs(prevSlope) + Math.abs(slope));
```

Comment:
same as above, I think (x - (x - step)) is just step

Reply:
Done